colexec: plumb through logical types into operators

yuzefovich opened the pull request:
**colexec: move the contents of two files into the templates**

Two structs were defined in separate files. Now their declaration has
been moved into the corresponding template files.

Release note: None

**colexec: unify templated function calls**

This commit reuses (where possible) existing makeFunctionRegex
utility function which makes a regular expression to search for
templated function calls and introduces another utility function to
generate the actual templated function call. The latter allows us to not
specify quotes in _tmpl files in those calls which is more pleasing to
the eyes.

Release note: None

**execgen: miscellaneous cleanups**

This commit removes some unused overloads from generation of
mergejoinbase and mergejoin_op templates. It also puts execgen
dependency into a template comment so that we don't need to have
a pointless statement to remove an unused warning.

Release note: None

**colexec: plumb through logical types into operators**

The "physical" type system (coltypes.T) doesn't contain all the needed
information in some cases, so this commit plumbs through the "logical"
types (sql/types.T) into the operators that might need that information.
Type-related arguments have been renamed to either be "logical" or
"physical" to highlight the distinction.

Addresses: #43559.

Release note: None

yuzefovich committed:
colexec: move the contents of two files into the templates

Two structs were defined in separate files. Now their declaration has
been moved into the corresponding template files.

Release note: None

yuzefovich committed:
colexec: unify templated function calls

This commit reuses (where possible) existing makeFunctionRegex
utility function which makes a regular expression to search for
templated function calls and introduces another utility function to
generate the actual templated function call. The latter allows us to not
specify quotes in _tmpl files in those calls which is more pleasing to
the eyes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43723)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: plumb through logical types into operators

The "physical" type system (coltypes.T) doesn't contain all the needed
information in some cases, so this commit plumbs through the "logical"
types (sql/types.T) into the operators that might need that information.
Type-related arguments have been renamed to either be "logical" or
"physical" to highlight the distinction.

Release note: None

yuzefovich committed:
execgen: miscellaneous cleanups

This commit removes some unused overloads from generation of
mergejoinbase and mergejoin_op templates. It also puts execgen
dependency into a template comment so that we don't need to have
a pointless statement to remove an unused warning.

Release note: None

yuzefovich commented:
Plumbing through the logical types as arguments to overloaded methods is not a good solution because we would need to do type checks within the tight loops. It seems to me that we will need to teach `overload` structs from `overloads.go` about logical types and somehow generate code based on a combination of logical and physical types.

At this point, I still think that this PR is valuable and gets us a step closer to solving type issues.

yuzefovich committed:
WIP on family + phys type generation

There is a problem of multiple IntFamily entries in the switch.

Release note: None

yuzefovich commented:
I extracted three commits into a separate PR and will postpone working on the colexec types refactor. I now think that we will need to introduce our own `colexec` type system, and I'm no longer sure that plumbing of the logical types is right move forward.
